DXMainClient/PreStartup.cs

```diff
@@ -73,6 +74,53 @@ public static void Initialize(StartupParams parameters)
 
             UserINISettings.Initialize(ClientConfiguration.Instance.SettingsIniName);
 
+            // Try to load translations
+            try
+            {
+                var translation = TranslationTable.LoadFromIniFile(ClientConfiguration.Instance.TranslationIniName);
+                LanguageInstance.TranslationTable = translation;
+                Logger.Log("Load translation: " + translation.LanguageName);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log("Failed to load the translation file. " + ex.Message);
+            }
+
+            try
+            {
+                if (ClientConfiguration.Instance.GenerateTranslationStub)
+                {
+                    // Not use using statement as variable "fs" will be destroyed after the lambda function be destroyed
+                    var fs = File.Open("Client/Translation.stub.ini", FileMode.Create);
+                    var sw = new StreamWriter(fs, new System.Text.UTF8Encoding(false)) { AutoFlush = true };
+                    Logger.Log("Generating translation stub feature is now enabled.");
+
+                    // Note, the TranslationTable.SaveIni() method is not used to avoid re-writing the file
```

Comment:
Also I am not sure why would you meed to duplicate code for that, couldn't you adjust the class methods for that?

Reply:
Calling `TranslationTable.SaveIni()` and append values to the end of the text stream requires the `[Translation]` section must be the last section in the ini file. Currently, the code does not guarantee that.

To solve the problem in this file the only way seems to be either re-writing the ini file everytime and sacrifice the performance, or generating the stub file only when program exits

Reply:
Is creating a timer for re-generating the ini file periodically acceptable?

Reply:
Is there any benefit to incrementally writing translation stubs instead of dumping them on unload? If yes then I guess we can keep this as is, if no - let's just dump the INI at exit.

Reply:
The benefit is to preserve the stub file if the program crashes... Well it seems indeed unnecessary.
Now I re-write this part to use a timer.
To dump the INI at exit, where is the currect location to write something where the program exit? Maybe after calling `new Startup().Execute()`?

Reply:
Well, I think maybe I can add a `ProgramExitEvent` to the program class and other components just subscribe to this event if they want to do something when exiting.

Reply:
That seems like a nice addition, I like the event way.

Reply:
`AppDomain.CurrentDomain.ProcessExit` is such an event. Now the stub file will be generated when program exits.